### DISCUSSION TITLE
NMI TOD clock alarm interrupt

### DISCUSSION THREAD

**Discussion by Whammo** [jkotlinski/durexforth#335] [Show and tell]
This is a number of applications to exploit the TOD clock. 
 The alarm, as it is a NMI, ~~calls QUIT, so it could be used as a watchdog timer.~~
  The alarm NMI now only gets cleared by the NMI routine as per Kernal.
Now ``ptod`` with colons!
Leverage the power of your c64 as a clock.

```forth
: demit dup emit ;

: bcdin key demit $30 -
>r 1 key demit dup
$d <> if $30 - >r 1+ key demit
     then emit
1- r> swap 0<> if r> $10 * or then ;

: ampm ." am or pm? "
key demit $41 <> if $80 or then cr ;

: hour ." 1-12? " bcdin ampm $dd0b c! ;
: min ." 0-59 " bcdin $dd0a c! ;
: sec ." 0-59 " bcdin $dd09 c! ;
: 10th ." 0-9 " bcdin $dd08 c! ;

: rtod $dd0b dup c@ dup $7f and >r
        $80 and swap
        1- dup c@ >r
        1- dup c@ >r
        1-     c@ >r
        r> r> r> r> ;

: ptod  hex rtod          \ prints TOD
3 0 do dup 10 u< if '0' emit then
. $9d emit ':' emit loop
. decimal $80 and if   ." pm"
                else ." am"
                then cr ;

: settod  \ sets TOD
hour min sec 10th ptod ;

: setalarm \ sets alarm, enables interr
$80 $dd0f c! \ alarm bit
settod $84 $dd0d c! \ nmi enable
$2a1 c@ 4 or $2a1 c! \ rs232 enables
0 $dd0f c! ;  \ tod bit

: clalarm 4 $dd0d c!
$2a1 c@ $fb and $2a1 c! ; \ disable nmi, rs232 enables

: elapsed 0 0 0 0 \ used for benchmarki
          $dd0b c!
          $dd0a c!
          $dd09 c!
          $dd08 c! ;

 : time source >in @ /string
dup >in +! elapsed evaluate
ptod ;
\ time $c000 dump

: clock page
begin $13 emit ptod key?
until key drop ;

**Comment by polluks** [comment]
Cool! There was a hardware bug at midnight IIRC...

**Reply by Whammo** [reply]
There is no correction for the defect that makes the TOD clocks follow:

11:59:59.9 AM
with
12:00:00.0 PM

instead of AM, although 0.1 seconds later, the clocks' PM will be correct.  A
similar flaw exists for midnight.

**Comment by Whammo** [comment]
An easy way to show this working is:
```forth
include tod
settod
setalarm
: go ptod recurse ;
go
``` 
~~It will break the GO loop accordingly.~~

**Comment by Whammo** [comment]
You'll notice when warp mode in VICE is active, unlike the jiffy clock, TOD isn't sped up.

**Comment by ekipan** [comment]
I rewrote your code to help myself understand it. It uses Johan's nice [`00.`](https://github.com/jkotlinski/durexforth/discussions/535#discussioncomment-5458694) word. Seems to work, though I know almost nothing about C64 hardware. The only reason `>b` isn't called `>bcd` is so `am` can fit on one line :P
```forth
decimal \ marker -tod- : vv -tod- v ;

\ time of day ( dsec sec min hour )
: >b  10 /mod 4 lshift or ; \ bcd
: am  >r 0 -rot r> >b swap >b rot >b ;
: pm  am $80 or ;
\ e.g. 4 30 57 pm -- 0 $57 $30 $84

\ print:
: 00.  s>d <# # # #> type ;
: a/p  $80 and if 'p' else 'a' then ;
: tod.  dup >r $1f and  base @ >r hex
 00. ':' emit 00. ':' emit 00. drop
 r> base !  r> a/p emit ." m " ;

\ fetch/store tod registers:
: c@+  dup c@ swap 1+ ;
: c!-  tuck c! 1- ;
: tod@  $dd08 c@+ c@+ c@+ c@ ;
: tod!  $dd0b c!- c!- c!- c! ;

\ alarm interrupt & register mode:
: alset!  if $84 else 4 then $dd0d c! ;
: almode! if $80 else 0 then $dd0f c! ;
: alarm!  1 almode! tod! 0 almode! ;

\ user interface:
: setalarm  1 alset! alarm! ;
: clalarm  0 alset! ;
: ptod  tod@ tod. ;
: lcsr  0 do $9d emit loop ;
: poll  key? if key else 0 then ;
: clock begin ptod 11 lcsr poll until ;
\ e.g. 1 23 45 pm tod! page clock

\ benching: e.g. ' slow timed
: slow  $a000 dump n n n ;
: timed  0 0 0 0 tod! execute ptod cr ;
: in> source >in @ /string dup >in +! ;
: time  in> ['] evaluate timed refill ;
```
Some history of this comment, check edits for full details:
- First edit reversed `tod!`'s input order, was actually broken before. Also inlined the Constants into `tod@` and `tod!`
- Then added a missing `drop` to `ptod`.
- Removed outdated wording and added this changelog.
- Extracted `timed` and `rest`, changed `keyin` to `chrin`
- Removed `chrin` `bcdin` `settod` added `am` `pm` `tod.` `alset!` `almode!` `alarm!`
- Deleted [incorrect](https://github.com/jkotlinski/durexforth/commit/eeca09a9911dfc5e32de89dde060036166d44200#r107742357) `chrin` from below. Changed `c@@`/`c!!` into `c@+`/`c!-`.
- `rest` into `in>`, as in Whammo's original code.

Here's the removed input words in case they might be useful elsewhere:
```forth
: keyin  key dup emit ;
: bcdin  chrin '0' - chrin '0' -
 dup 9 u> if drop exit then
 swap 4 lshift or ;
```

**Reply by ekipan** [reply]
It occurs to me that `tod@` and `tod!` aren't inverses.
```
1 2 3 4 tod! tod@ .s tod! tod@ .s
1 2 3 4 4 3 2 1 ok
```
Editing to fix that problem and also make a bit more concise.

**Reply by ekipan** [reply]
And now settod is setting base to 0 somewhere, causing it to crash. I'll mull this over a bit.

Edit: found it! Forgot to drop the deciseconds in `ptod`. `clock` would blow the stack after a couple seconds.

**Reply by Whammo** [reply]
You are fortunate. Poking around in such things is perhaps less fun in Assembly, and certainly less fun with BASIC. 👍

**Reply by ekipan** [reply]
Speaking of assembly, here's a couple more niceties. A more general `time` and `chrin` instead of `keyin`, with perks like editing and a cursor.

**Edit 1:** though for some reason `timed` always prints 00:00:00am, no matter how long the executed word takes. Not sure what's going on there.
**Edit 2:** `time $a000 dump n n n` gave four seconds. Guess one dump wasn't slow enough :P
**Edit 3:** moved the code to the main post.

**Reply by ekipan** [reply]
I hope it's not a bother if I post more pickings at this.

I wrote `am` and `pm` words that convert hour minute second into bcd and rearrange them into the format `tod!` and `tod.` expect (also extracted `tod.` from `ptod`). That way you can just reuse the forth interpreter and delete the number input words. 

**Edit:** Code moved into main post.

The alarm doesn't seem to do anything, unless I'm supposed to write more code that responds to it somehow, not sure.
```forth
1 23 50 pm setalarm
ok
1 23 45 pm tod! clock
ok:23:52pm
```

**Reply by Whammo** [reply]
Not a bother at all. In fact, it's why I bother to post my own musings.

**Reply by Whammo** [reply]
Alarm is a matter of pointing the NMI vector to one's own assembly code.

**Comment by Whammo** [comment]
restore_handler in interpreter.asm is mostly a copy of the kernal's NMI handler.
 This is easily durexForthified and and appended with the appropriate code to redirect it to your favorite word when appropriate and to the kernal's routines or ```quit``` when it is appropriate.
```asm
restore_handler
    pha             ; save a
    txa             ; copy x
    pha             ; save x
    tya             ; copy y
    pha             ; save y
    lda	#$7f        ; disable all CIA 2 interrupts
    sta	$dd0d       ;
    ldy	$dd0d       ; save CIA 2 interrupt control register for kernal_nmi
    bpl brk_handler ; CIA 2 is not the NMI source if the most significant bit is not set.

kernal_nmi
    jmp $fe72       ; all CIA 2 NMI's fall through to the Kernals' RS-232 routines

brk_handler         ; all non-CIA NMI (RESTORE key) and brk instructions- via IRQ vector end up here.
    pla             ; drop y -the return stack will be reset by QUIT anyway
    pla             ; pull x
    tax             ; restore parameter stack pointer for QUIT
    jmp QUIT        ; already under sei from NMI stub in Kernal or from IRQ to brk_handler
```
such as this, as yet, untested:
Proof that labels are cool.
Bit 2 needs to be set in $2a1 ENABL or rs232 activity will clear the CIA alarm NMI enable bit.
 

```forth
code rsthandler 
pha, txa, pha, tya, pha, 
$7f lda,# $ddod dup sta, ldy,
0 @: bpl,
tya, 4 and,# 1 @: bne,
$fe72 jmp, 0 @@
pla, pla, tax, ' quit jmp,
1 @@  ' mycode jsr, \ mind yr, $fe72 requires it.
$fe72 jmp,  \ let the Kernal deal with the stack
;code

```
 We have the NMI code. Now we want to signal an alarm without causing issues with the console, or another running program.
If we use the opportunity of the NMI to change the IRQ vector to our signal, say, flashing the border color on the IRQ and playing a ```bell``` sound every 60 jiffies we will presumably also require yet another NMI routine to expect RESTORE in order to clear the alarm by resetting the IRQ vector.
